src/utils.d.ts

```diff
@@ -0,0 +1,5 @@
+type UndefinedProperties<T> = {
```

Comment:
Can you please add a comment to the file and explain here how this works.

Looks like utter magic to me. To support this, I need to understand how it works.

Reply:
Similar code to this can be found here. in line 180 utility PickByValue

https://github.com/piotrwitek/utility-types/blob/master/src/mapped-types.ts

Reply:
That is an awesome resource.

We definitely want zero dependencies. So just a ref here is fine. Is see it is MIT licensed so that is okay to learn from it and adapt. 

Can you add comments to the code to do your best to explain so the code can be supported by others too?

Reply:
I already did. Let me know if it's not sufficient.

Reply:
Cool, sorry I missed that. 

Looks great. I'll merge later today.

Reply:
Absolutely. Thank you!